routing/turns_generator.cpp

```diff
@@ -373,12 +380,103 @@ size_t GetLinkCount(vector<TurnCandidate> const & candidates)
+                                   NumMwmIds const & numMwmIds, RoutePointIndex & nextIndex)
+{
+  auto const & segments = result.GetSegments();
+  ASSERT_LESS(index.m_segmentIndex, segments.size(), ());
+  ASSERT_LESS(index.m_pathIndex, segments[index.m_segmentIndex].m_path.size(), ());
+
+  if (index.m_pathIndex + 1 != segments[index.m_segmentIndex].m_path.size())
+  {
+    // In segment case.
+    nextIndex = {index.m_segmentIndex, index.m_pathIndex + 1};
+    return true;
+  }
+
+  // Case when the last point of the current segment is reached.
+  // So probably it's necessary to cross a segment border.
+  if (index.m_segmentIndex + 1 == segments.size())
+    return false; // The end of the route is reached.
+
+  TurnInfo const turnInfo(segments[index.m_segmentIndex], segments[index.m_segmentIndex + 1]);
+  ASSERT_GREATER_OR_EQUAL(turnInfo.m_ingoing.m_path.size(), 2, ());
+  ASSERT_GREATER_OR_EQUAL(turnInfo.m_outgoing.m_path.size(), 2, ());
+
+  double const oneSegmentTurnAngle = my::RadToDeg(
+      PiMinusTwoVectorsAngle(turnInfo.m_ingoing.m_path.back().GetPoint(),
+                             turnInfo.m_ingoing.m_path[turnInfo.m_ingoing.m_path.size() - 2].GetPoint(),
+                             turnInfo.m_outgoing.m_path[1].GetPoint()));
+  CarDirection const oneSegmentDirection = IntermediateDirection(oneSegmentTurnAngle);
+  if (!IsGoStraightOrSlightTurn(oneSegmentDirection))
+    return false; // Too sharp turn angle.
+
+  size_t ingoingCount = 0;
+  TurnCandidates possibleTurns;
+  result.GetPossibleTurns(turnInfo.m_ingoing.m_segmentRange, GetPointByIndex(segments, index),
+                          ingoingCount, possibleTurns);
+
+  if (possibleTurns.candidates.empty())
+    return false;
+
+  // |segments| is a vector of |LoadedPathSegment|. Every |LoadedPathSegment::m_path|
+  // contains junctions of the segment. The first junction at a |LoadedPathSegment::m_path|
+  // is the same (or almost the same) with the last junction at the next |LoadedPathSegment::m_path|.
+  // To prevent using the same point twice it's necessary to take the first point only from the
+  // first item of |loadedSegments|. The beginning should be ignored for the rest |m_path|.
+  // Please see a comment in MakeTurnAnnotation() for more details.
+  if (possibleTurns.candidates.size() == 1)
+  {
+    // Taking the next point of the next segment.
+    nextIndex = {index.m_segmentIndex + 1, 1 /* m_pathIndex */};
+    return true;
+  }
+
+  if (!KeepTurnByHighwayClass(possibleTurns, turnInfo, numMwmIds))
+  {
+    // Taking the next point of the next segment.
+    nextIndex = {index.m_segmentIndex + 1, 1 /* m_pathIndex */};
+    return true;
+  }
+  // Stopping getting next route points because an important bifurcation point is reached.
+  return false;
+}
+
+bool GetPrevInSegmentRoutePoint(RoutePointIndex const & index, RoutePointIndex & nextIndex)
+{
+  if (index.m_pathIndex == 0)
```

Comment:
почему мы тут не перескакиваем в конец предыдущего сегмента, если он есть?

Reply:
Обрати п-та внимание, на комент в hpp  файле:
```
+/*!
+ * \brief Calculates |nextIndex| which is an index of next route point at result.GetSegments().
+ * |nextIndex| may be calculated in forward and backward direction.
+ * \returns true if |nextIndex| is calculated and false otherwise. This method returns false
+ * if looking for next index should be stopped. For example in case of the point before first,
+ * after the last in the route and sometimes at the end of features and at feature intersections.
+ * \note This method is asymmetric. It means that in some cases it's possible to cross
+ * the border between turn segments in forward direction, but in backward direction (|forward| == false)
+ * it's impossible. So when turn segment border is reached in backward direction the method
+ * returns false. The reasons why crossing the border in backward direction is not implemented is
+ * it's almost not used now and it's difficult to cover it with unit and integration tests.
+ * As a consequence moving in backward direction it's possible to get index of the first item
+ * of a segment. But it's impossible moving in forward direction.
+ */
+bool GetNextRoutePointIndex(IRoutingResult const & result, RoutePointIndex const & index,
+                            NumMwmIds const & numMwmIds, bool forward, RoutePointIndex & nextIndex);
```
Он как раз про это. Если коротко - я это не написал, поскольку это очень непросто протестировать и сейчас, не очень используется в коде (Т.е. использоваться будет, но не значительно).